javascript/treeIntersection/tree-intersection.js

```diff
@@ -0,0 +1,106 @@
+'use strict';
+
+class Node {
+    constructor(value) {
+      this.value = value;
+      this.left = null;
+      this.right = null;
+    }
+  }
+  
+  class BinarySearchTree {
+    constructor() {
+      this.root = null;
+    }
+  
+    insert(value) {
+      const newNode = new Node(value);
+  
+      if (!this.root) {
+        this.root = newNode;
+        return;
+      }
+  
+      let current = this.root;
+  
+      while (true) {
+        if (value === current.value) {
+          return;
+        }
+  
+        if (value < current.value) {
+          if (!current.left) {
+            current.left = newNode;
+            return;
+          }
+          current = current.left;
+        } else {
+          if (!current.right) {
+            current.right = newNode;
+            return;
+          }
+          current = current.right;
+        }
+      }
+    }
+  
+    // Helper function for tree traversal
+    _traverse(node, callback) {
+      if (node) {
+        this._traverse(node.left, callback);
+        callback(node.value);
+        this._traverse(node.right, callback);
+      }
+    }
+  
+    // Returns an array of values in the tree
+    traverse() {
+      const values = [];
+      this._traverse(this.root, value => values.push(value));
+      return values;
+    }
+  }
+  
+  function tree_intersection(tree1, tree2) {
+    const hashmap = new Map();
+    const commonValues = new Set();
+  
+    // Traverse the first tree and add values to the hashmap
+    tree1.traverse().forEach(value => {
+      hashmap.set(value, true);
+    });
```

Comment:
By traversing and creating array, you are adding unnecessary time and space with an extra step and another new data structure. Instead, take advantage of the helper function you have in your class:

``` js
tree1._traverse(tree1.root, () => hashmap.set(value, true));